0000-ilp-over-http.md

```diff
@@ -0,0 +1,81 @@
+---
+draft: 1
+title: ILP Over HTTP
+---
+# ILP Over HTTP
+
+This spec defines a minimal [Ledger Layer](./0001-interledger-architecture/0001-interledger-architecture.md#ledger-layer) protocol using the Hyper Text Transfer Protocol (HTTP). It includes the Interledger payment fields (such as destination address) alongside those that are local to a specific hop (such as the transfer amount).
+
+**Note:** This protocol is not to be confused with [HTTP ILP](./0014-http-ilp/0014-http-ilp.md), which defines HTTP headers for communicating payment details on a normal HTTP request to an application server. The values defined in this protocol will be forwarded to the destination ILP address defined in the request.
+
+## Overview
+
+It's everything you need to do ILP with just a couple of HTTP headers on a POST request/response. What more do we need to say?
+
+## Specification
+
+This protocol SHOULD be used with HTTPS or HTTP/2. It SHOULD NOT be used with insecure HTTP except for testing purposes.
+
+### Request
+
+```http
+POST / HTTP/1.1
+ILP-Destination: g.crypto.bitcoin.1XPTgDRhN8RFnzniWCddobD9iKZatrvH4.~asdf1234
+ILP-Condition: x73kz0AGyqYqhw/c5LqMhSgpcOLF3rBS8GdR52hLpB8=
+ILP-Expiry: 2017-12-07T18:47:59.015Z
+ILP-Amount: 1000
+
+<body>
+```
+
+| Field | Type | Modified at Each Hop? | Description |
+|---|---|---|---|
+| `ILP-Destination` | [ILP Address](./0015-ilp-addresses/0015-ilp-addresses.md) | N | Destination address of the payment |
+| `ILP-Condition` | Base64-Encoded String, 32 Bytes | N | Execution condition of the payment, which is the Sha256 hash digest of the condition fulfillment |
+| `ILP-Expiry` | [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) Timestamp in UTC | Y | Expiry of the transfer |
+| `ILP-Amount` | Unsigned 64-Bit Integer | Y | Transfer amount, denominated in the minimum divisible units of the ledger. Note that this is the local transfer amount, **not** the destination amount as in the original [ILP Payment Packet Format](https://github.com/interledger/rfcs/blob/master/0003-interledger-protocol/0003-interledger-protocol.md#ilp-payment-packet-format) |
+| `<body>` | Binary, Maximum of 32767 Bytes | N | End-to-end data used by Transport Layer protocols |
+| `ILP-...` | UTF-8 String | N | Additional headers that MUST be forwarded by all connectors, even if they are not understood |
```

Comment:
Don't think this is necessary. anything that goes in the headers is ledger layer data and by implication the server should understand it.

Reply:
Regarding @adrianhopebailie's comment:
> If there is data that needs to be forwarded blindly by the connector then it should be in the body.

Right now the body is intended for communication between the sender and receiver. What's missing from that is any kind of information the sender may want to communicate to the connectors or that connectors may want to attach to the payment. If transport layer protocols don't have a way to tell the receiver to ignore some portion of the data, attaching additional information would cause the payment to fail if the receiver HMACs the data to generate the fulfillment.

I think the questions we need to answer about the type of extensibility we want are:
1. Who is going to set the additional fields? Will they be supported and set by Transport Layer protocols, by the sender's ILP implementation, or by connectors?
2. Are we looking for the type of extensibility where ILP's IETF would define new fields on the Interledger layer, or something more open like HTTP headers where anyone can define new ones?
3. How do we ensure that the new features are forwarded even by connectors that don't understand them (which would allow for rolling upgrades, rather than requiring everyone to upgrade to use the new feature at once)?

Reply:
In my opinion we are falling into the trap of mixing protocol layers again.

> Right now the body is intended for communication between the sender and receiver. 

That's not strictly speaking true. The body is intended for use by any protocol in the stack above ILP.

If a transport layer protocol defines data that is only for the senders and receivers but also data that must be visible to the intermediaries then it should define an envelop that is carried in the body of the ledger layer message and distinguishes between these.

PSK does this already as one example.

I'd consider any transport layer that doesn't provide a way to pass both public and protected (encrypted by sender/receiver) data as a non-starter anyway.

Consider a transport layer protocol that needs to provide compliance data to each intermediary. So each connector attaches a whole lot of data to the transfer on each hop that describes who they are and a signature etc.

There is no way you want that passed in the headers of the ledger layer protocol.

To answer your questions:

1. Protocols don't set fields. They define which fields can be set by who. A good transport layer protocol will define both fields that can be set by intermediaries and by the sender and receiver but these will all be in the body of the ledger layer message.

2. Yes. The ILP layer is very rigid and is unlikely to change again. It carries a local transfer amount, an address, a condition, a timeout and the transport layer data. There is nothing left to remove and very unlikely to be anything more to add that shouldn't be in the transport layer.

3. Put them in the body of the transfer message.

Reply:
I agree that transport protocols that want to expose data for intermediaries to see can do so.

I think this is what @justmoon is worried about:
> The ILP layer is very rigid...and very unlikely to be anything more to add that shouldn't be in the transport layer.

It comes down to whether the Interledger layer itself should have extensibility. We can add extensibility at the ledger and transport layers easily, but if there's a field we realized we forgot or desperately need at the Interledger layer it would be a huge pain or near impossibility to add later.

Reply:
> It comes down to whether the Interledger layer itself should have extensibility

I agree and I think the answer is an emphatic no. The inter-networking layer needs to be simple and never change.

Reply:
> A good transport layer protocol will define both fields that can be set by intermediaries and by the sender and receiver but these will all be in the body of the ledger layer message.

I don't think this spec should be prescriptive about what goes into the body of the HTTP request. For example, imagine some sender/receiver that don't want to stuff all of their payment data into this layer at all, but instead simply want to hash it, encrypt it, and then propagate a link to that data? 

This protocol should be able to allow a sender and receiver to do that without involving the `<body>` at all. As an example, to satisfy the above, one could define a URL-link to shared data that might be too large to send across every hop, resulting in an HTTP request that looks like this:

```
{
POST / HTTP/1.1
ILP-Destination: g.crypto.bitcoin.1XPTgDRhN8RFnzniWCddobD9iKZatrvH4.~asdf1234
ILP-Condition: x73kz0AGyqYqhw/c5LqMhSgpcOLF3rBS8GdR52hLpB8=
ILP-Expiry: 2017-12-07T18:47:59.015Z
ILP-Amount: 1000
ILP-SenderReceiver-Encrypted-Data: https://s3.exmple.com?id=32e77ef1961f1c9c5135bff6e939b7ef3b4e8c7defa61d50dca797c86f97cc91
}
```

Reply:
> It comes down to whether the Interledger layer itself should have extensibility. We can add extensibility at the ledger and transport layers easily, but if there's a field we realized we forgot or desperately need at the Interledger layer it would be a huge pain or near impossibility to add later.

It seems like all we need is an extensible way to allow intermediate hops (per this spec) to be able to identify how to handle headers according to the following:

1. Which header's data do I need to adjust, and then forward? (e.g., `Amount`, `expiresAt`)
1. Which header's data MUST I forward unchanged? (e.g., `destination`, `executionCondition`)
1. Which header's data SHOULD I forward unchanged? 
1. Which header's data is meant only for me?

Is seems like we could use a combination of header-prefixes and/or this spec to cover all of the above cases, _and_ allow for extensibility. 

Maybe a simple rule would be: Read the spec, do what it says, and then: for any _other_ field you encounter that starts with `ILP-...`, you MUST forward to any next-hop, unchanged.